iothub/service/src/Exceptions/ExceptionHandlingHelper.cs

```diff
@@ -18,181 +16,131 @@ internal class ExceptionHandlingHelper
 
-                    if (messageFields != null)
+            if (messageFields == null && Logging.IsEnabled)
+            {
+                Logging.Error(
+                    nameof(GetIotHubErrorCodeAsync),
+                    $"Failed to find expected semicolon in error message to find error code." +
+                    $" Message body: '{responseBody}.'");
+            }
+            else
+            {
+                foreach (string messageField in messageFields)
+                {
+                    if (messageField.IndexOf(MessageFieldErrorCode, StringComparison.OrdinalIgnoreCase) >= 0)
                     {
-                        foreach (string messageField in messageFields)
+                        const char errorCodeDelimiter = ':';
+
+                        if (messageField.IndexOf(errorCodeDelimiter) >= 0)
                         {
-                            if (messageField.IndexOf(MessageFieldErrorCode, StringComparison.OrdinalIgnoreCase) >= 0)
+                            string[] errorCodeFields = messageField.Split(errorCodeDelimiter);
+
+                            string returnedErrorCode = errorCodeFields[1];
+
+                            // When the returned error code is numeric, only take the first 6 characters as it contains 6 digits.
+                            if (int.TryParse(returnedErrorCode.Substring(0, 6), out int code))
                             {
-                                const char errorCodeDelimiter = ':';
-
-                                if (messageField.IndexOf(errorCodeDelimiter) >= 0)
-                                {
-                                    string[] errorCodeFields = messageField.Split(errorCodeDelimiter);
-                                    if (Enum.TryParse(errorCodeFields[1], out ErrorCode errorCode))
-                                    {
-                                        errorCodeValue = (int)errorCode;
-                                    }
-                                }
+                                return CompareToResponseHeader(response, (IotHubErrorCode)code);
                             }
-                            break;
-                        }
-                    }
-                    else
-                    {
-                        if (Logging.IsEnabled)
-                            Logging.Error(null, $"Failed to deserialize error message into a dictionary and could not parse ';' delimited string either: {ex}." +
-                                $" Message body: '{responseContentStr}.'");
 
-                        return ErrorCode.InvalidErrorCode;
+                            // Otherwise the error code might be a string (e.g., PreconditionFailed) in which case we'll try to
+                            // find the matching IotHubErrorCode enum with that same name.
+                            if (Enum.TryParse(returnedErrorCode, out IotHubErrorCode errorCode))
```

Comment:
I feel like this code block should be outside of the message fields foreach. We should only do this if that attempt to find a best match fails.

Reply:
This code block is for the case non-numeric error codes in the response content (i.e., "PreconditionFailed"), so I prefer to leave it as-is. If both attempts of matching numeric and non-numeric codes fail, we will simply return `IotHubErrorCode.Unknown`.